src/client/datascience/notebook/vscodeNotebookController.ts

```diff
@@ -377,15 +369,31 @@ export class VSCodeNotebookController implements Disposable {
             .then(noop, (ex) => console.error(ex));
     }
 
-    private executeCell(doc: NotebookDocument, cell: NotebookCell) {
+    private async executeCell(doc: NotebookDocument, cell: NotebookCell) {
         traceInfo(`Execute Cell ${cell.index} ${getDisplayPath(cell.notebook.uri)}`);
-        const kernel = this.kernelProvider.getOrCreate(cell.notebook, {
-            metadata: this.kernelConnection,
-            controller: this.controller,
-            resourceUri: doc.uri
-        });
-        this.updateKernelInfoInNotebookWhenAvailable(kernel, doc);
-        return kernel.executeCell(cell);
+
+        // Start execution now (from the user's point of view)
+        const execution = CellExecutionCreator.getOrCreate(cell, this.controller);
+        execution.start(new Date().getTime());
+        void execution.clearOutput(cell);
+
+        // Connect to a matching kernel if possible (but user may pick a different one)
+        try {
+            const kernel = await this.connectToKernel(doc);
+            this.updateKernelInfoInNotebookWhenAvailable(kernel, doc);
+            return await kernel.executeCell(cell);
+        } catch (ex) {
+            // If there was a failure connecting or executing the kernel, stick it in this cell
+            return displayErrorsInCell(cell, execution, ex.toString());
```

Comment:
This doesn't look good.
We're just dumpgin a call stack.
I"d like us to analyse the error and provide better information instead of just using a toString

Reply:
I don't think we had any code that dumped the raw errors like this, this is new code and would be a poor experience

Reply:
E.g. if the kernel dies half way, then dumping a call stack in the output is a poor ux, and that's one of the things we tried to fix. I don't think we should be displaying node.js stack trace at all.

Reply:
I don't think ```ex.toString()``` includes the callstack.

I wrote the following code:

```javascript
function throwStuff() {
   throw new Error('Here is an error');
}

function doStuff() {
   try {
      throwStuff();
   }
   catch (ex) {
      console.log(ex.toString());
      console.log('Here is the stack');
      console.log(ex.stack);
   }
}

doStuff();
```

It output this:

```
Error: Here is an error
Here is the stack
Error: Here is an error
    at throwStuff (C:\Users\aku91\test.js:2:10)
    at doStuff (C:\Users\aku91\test.js:7:7)
    at Object.<anonymous> (C:\Users\aku91\test.js:16:1)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at internal/main/run_main_module.js:17:47
```

So the callstack only comes out when asked for.